### PR TITLE
Piecewise fix for 0d

### PR DESCRIPTION
**Pull request by astrojuanlu**
I provide a fix for #331. Actually little changes were needed, but I had to carefully think about the assumptions being made by the function. Some additional tests were added and the behavior remains backwards compatible.

Besides, I corrected the documentation because there was an inconsistent bit, see 0d1f7d3.

### CODE REVIEW

**Comment by jakobgager** [comment]
There, seems to be an issue with scalar domains and three or more conditions:
```python
>>> np.piecewise(3, [True, False, False],[4,2,0])
ValueError: function list and condition list must be the same    
```
Funny enough it works fine with two conditions (this is what has been tested):
```python
>>> np.piecewise(3, [True, False],[4,2])
array(4)
```

The three conditions version works fine with _select_:
```python
>>> np.select([True, False, False],[4,2,0])
array(4)
```

**Comment by Juanlu001** [comment]
I won't be able to fix this for some time, I fear :(

**Comment by bastula** [comment]
The changes applied from this pull request don't seem to work well with multidimensional arrays that have an extra "otherwise" condition:

```python
>>> x = array([[-2.5, -1.5, -0.5],
               [ 0.5,  1.5,  2.5]])
>>> np.piecewise(x, [x < 0, x >= 2], [-1, 1, 3])
ValueError: all the input arrays must have same number of dimensions
```

The same code produces the following result in numpy 1.8.2:
```python
>>> array([[-1., -1., -1.],
           [ 3.,  3.,  1.]])
```

If you reshape the array to a single dimension, it works as intended.

Also, I noticed there aren't any multidimensional test cases, so maybe this can be one of them?

**Comment by jaimefrio** [comment]
@jakobgager, @bastula Can you convert your comments here into issues? It is the only way to not lose track of them.

File: numpy/lib/tests/test_function_base.py [hunk 1]
```diff
@@ -1505,6 +1506,15 @@ def test_0d(self):
         assert_(y.ndim == 0)
         assert_(y == 0)
 
+        x = 5
+        y = piecewise(x, [[True], [False]], [1, 0])
```
Comment (eric-wieser) [inline numpy/lib/tests/test_function_base.py]: This doesn't make any sense. If `x` is 0d, the conditions should be 0d too, as `[True, False]`

**Pull request merged**